File: inferlet-js/src/runtime.ts
// Runtime wrapper functions for inferlet-core-runtime bindings
// These functions wrap the WASM component bindings with a clean TypeScript API

import * as runtime from 'inferlet:core/runtime';
import { awaitFuture } from './async-utils.js';

/**
 * Returns the runtime version string
 */
export function getVersion(): string {
  return runtime.getVersion();
}

/**
 * Returns a unique identifier for the running instance
 */
export function getInstanceId(): string {
  return runtime.getInstanceId();
}

/**
 * Parses an array of command line arguments into a JSON object.
 * 
 * Rules:
 * 1. Flags starting with '-' or '--' become keys.
 * 2. If a flag is followed by a value (not starting with '-'), it's a key-value pair.
 * 3. If a flag is followed by another flag or nothing, it is treated as boolean true.
 */
function parseArgs(args: string[]): Record<string, string | boolean> {
  const parsed: Record<string, string | boolean> = {};

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];

    // Only process flags
    if (arg.startsWith('-')) {
      // Remove leading dashes
      const key = arg.replace(/^-+/, '');

      // Peek at the next argument
      const nextArg = args[i + 1];

      // If next arg exists and is NOT a flag, it is the value
      if (nextArg && !nextArg.startsWith('-')) {
        parsed[key] = nextArg;
        i++; // Skip the next index as we just consumed it
      } else {
        // Otherwise, treat it as a boolean flag
        parsed[key] = true;
      }
    }
  }

  return parsed;
}

/**
 * Retrieves POSIX-style CLI arguments passed to the inferlet from the remote user client,
 * parsed into an object with key-value pairs.
 */
export function getArguments(): Record<string, string | boolean> {
  return parseArgs(runtime.getArguments());
}

/**
 * Sets the return value for the inferlet
 */
export function setReturn(value: string): void {
  runtime.setReturn(value);
}

/**
 * Executes a debug command and returns the result as a string (async)
 * This function handles the pollable resource to wait for the result
 */
export async function debugQuery(query: string): Promise<string> {
  return awaitFuture(runtime.debugQuery(query), 'debugQuery result was undefined');
}

File: sdk/python/src/inferlet/runtime.py
"""
Runtime wrapper functions for inferlet-core-runtime bindings.
Provides Pythonic API over WIT bindings.
"""

from typing import Any

from wit_world.imports import runtime as _runtime
from .async_utils import await_future

# Track whether set_return has been called
_return_was_set = False


def get_version() -> str:
    """Returns the runtime version string."""
    return _runtime.get_version()


def get_instance_id() -> str:
    """Returns a unique identifier for the running instance."""
    return _runtime.get_instance_id()


def _parse_args(args: list[str]) -> dict[str, str | bool]:
    """
    Parse POSIX-style CLI arguments into a dict.

    Rules:
    1. Flags starting with '-' or '--' become keys.
    2. If a flag is followed by a value (not starting with '-'), it's a key-value pair.
    3. If a flag is followed by another flag or nothing, it's treated as boolean True.
    """
    parsed: dict[str, str | bool] = {}
    i = 0

    while i < len(args):
        arg = args[i]

        if arg.startswith("-"):
            # Remove leading dashes
            key = arg.lstrip("-")

            # Peek at next argument
            next_arg = args[i + 1] if i + 1 < len(args) else None

            # If next arg exists and is NOT a flag, it's the value
            if next_arg and not next_arg.startswith("-"):
                parsed[key] = next_arg
                i += 2
            else:
                # Boolean flag
                parsed[key] = True
                i += 1
        else:
            i += 1

    return parsed


def get_arguments() -> dict[str, Any]:
    """
    Retrieves CLI arguments passed to the inferlet, parsed into a dict.

    Example:
        '--prompt hello --verbose' -> {'prompt': 'hello', 'verbose': True}
    """
    raw_args = _runtime.get_arguments()
    return _parse_args(raw_args)


def set_return(value: str) -> None:
    """Sets the return value for the inferlet."""
    global _return_was_set
    _return_was_set = True
    _runtime.set_return(value)


def was_return_set() -> bool:
    """Check if set_return was called."""
    return _return_was_set


def debug_query(query: str) -> str:
    """
    Executes a debug command and returns the result as a string.

    This is a module-level function that provides debug query capabilities
    without requiring a Queue instance.

    Args:
        query: The debug query string to execute.

    Returns:
        The result of the debug query as a string.
    """
    result = _runtime.debug_query(query)
    return await_future(result, "debug_query result was None")
